### ISSUE TITLE
Model returns all-zero predictions

### ISSUE THREAD

**Issue by sainatarajan** [mazurowski-lab/mri_foundation#7]
I am attempting to use your model for a knee MRI bone segmentation task, which involves identifying the femur, tibia, patella, and fibula from 3D MRI volumes. My current issue is that when I use the model directly for inference, it runs successfully but produces an output volume consisting entirely of zeros, effectively failing to segment any part of the image. 

I was unable to locate sufficient help or relevant examples for inference on an MRI and save its segmentation output. So I proceeded to create a custom script to handle the inference process:

```
def create_args():
    """Create args matching the notebook - hardcoded values"""
    args = Namespace()
    # From user specifications
    args.arch = 'vit_b'
    args.num_cls = 1
    
    # Standard values from the notebook/repo
    args.net = 'sam'
    args.image_size = 1024
    args.out_size = 256
    args.if_encoder_adapter = True
    args.if_mask_decoder_adapter = True
    args.encoder_adapter_depths = [0, 1, 10, 11]
    args.decoder_adapt_depth = 2
    args.if_encoder_lora_layer = False
    args.if_decoder_lora_layer = False
    args.encoder_lora_layer = []
    args.if_update_encoder = True
    
    # Additional args that might be needed
    args.baseline = 'unet'
    args.model = 'ours'
    args.finetune_type = 'adapter'
    args.normalize_type = 'sam'
    args.dataset_name = 'inference'
    args.targets = 'combine_all'
    args.cls = -1
    
    return args

def preprocess_slice(slice_2d, target_size=1024, use_sam_norm=False):
    """
    Float32 preprocessing:
     - per-slice min-max to [0,1] if use_sam_norm=False (medsam)
     - else min-max -> [0,1] then ImageNet mean/std (SAM)
     - returns torch.FloatTensor [3, H', W']
    """
    assert slice_2d.ndim == 2
    s = slice_2d.astype(np.float32)
    smin, smax = float(s.min()), float(s.max())
    if smax > smin:
        s = (s - smin) / (smax - smin)
    else:
        s = np.zeros_like(s, dtype=np.float32)

    # stack to 3 channels
    arr = np.stack([s, s, s], axis=0)  # (3, H, W)
    tensor = torch.from_numpy(arr).float()  # [3,H,W]

    # optional SAM ImageNet norm
    if use_sam_norm:
        mean = torch.tensor([0.485, 0.456, 0.406]).view(3,1,1)
        std  = torch.tensor([0.229, 0.224, 0.225]).view(3,1,1)
        # ensure same device later in forward
        tensor = (tensor - mean) / std

    return tensor  # [3, H', W']

@torch.no_grad()
def segment_slice(model, img_tensor, original_size, device, threshold=0.5):
    """
    Produces a binary mask numpy uint8 of shape original_size (W,H).
    - img_tensor: [3,H,W] float already normalized
    - original_size: PIL-style (width,height) returned by Image.size previously; convert to (H,W)
    """
    img_batch = img_tensor.unsqueeze(0).to(device)  # [1,3,H,W]

    # image embeddings
    img_emb = model.image_encoder(img_batch)

    # get prompt encodings (no prompts)
    sparse_emb, dense_emb = model.prompt_encoder(points=None, boxes=None, masks=None)
    # optionally get dense positional encodings if available
    image_pe = None
    if hasattr(model.prompt_encoder, 'get_dense_pe'):
        try:
            image_pe = model.prompt_encoder.get_dense_pe()
        except Exception:
            image_pe = None

    # mask decoder
    pred, iou_pred = model.mask_decoder(
        image_embeddings=img_emb,
        image_pe=image_pe,
        sparse_prompt_embeddings=sparse_emb,
        dense_prompt_embeddings=dense_emb,
        multimask_output=True,
    )  # pred: [1, num_masks, H_out, W_out]

    # choose best mask
    if iou_pred is not None:
        best_idx = int(torch.argmax(iou_pred[0]).cpu().item())
    else:
        best_idx = 0

    logits = pred[0, best_idx:best_idx+1]  # [1, H_out, W_out]

    # upsample logits to original image size (original_size is (W,H))
    orig_W, orig_H = original_size
    logits_up = F.interpolate(logits.unsqueeze(0), size=(orig_H, orig_W), mode='bilinear', align_corners=False).squeeze(0).squeeze(0)

    probs = torch.sigmoid(logits_up)
    mask = (probs > threshold).cpu().numpy().astype(np.uint8)  # H x W

    return mask

def process_volume(nii_path, model, device, image_size,
                   use_sam_norm=False, threshold=0.5,
                   save_overlays=False, overlay_dir=None):
    """
    Process a single NIfTI volume and return (seg_vol, affine, header).
    """
    nii = nib.load(str(nii_path))
    vol = nii.get_fdata()  # shape assumed (H, W, D)
    if vol.ndim != 3:
        raise ValueError(f"Expected 3D volume; got {vol.shape}")

    H, W, n_slices = vol.shape
    seg_vol = np.zeros((H, W, n_slices), dtype=np.uint8)

    if save_overlays and overlay_dir is not None:
        overlay_dir = Path(overlay_dir)
        overlay_dir.mkdir(parents=True, exist_ok=True)

    print(f"{nii_path.name}\n  Shape: ({H}, {W}, {n_slices})\n  Processing {n_slices} slices...")

    for i in tqdm(range(n_slices), desc="Slices"):
        slice_2d = vol[:, :, i]
        if np.all(slice_2d == 0):
            continue

        # original_size expected by segment_slice is (W, H)
        original_size = (W, H)

        # preprocess (keep float, optional sam norm, resize to model image_size)
        img_tensor = preprocess_slice(slice_2d, target_size=image_size, use_sam_norm=use_sam_norm)

        # segment_slice expects (model, img_tensor, original_size, device, threshold)
        mask_slice = segment_slice(model, img_tensor, original_size, device, threshold=threshold)

        # ensure shape matches (H, W)
        if mask_slice.shape != (H, W):
            # defensive fallback: try transpose if swapped
            if mask_slice.shape == (W, H):
                mask_slice = mask_slice.T
            else:
                raise RuntimeError(f"segment_slice returned shape {mask_slice.shape} but expected {(H, W)}")

        seg_vol[:, :, i] = mask_slice.astype(np.uint8)

    return seg_vol, nii.affine, nii.header
```

Can you please let me know if I am doing everything correctly?

**Comment by haoyudong-97** [comment]
Hi, your implementation looks correct. The reason the predictions are all zeros is that only the image encoder was pre-trained. Without a properly trained connection to the mask decoder, initializing it with SAM’s weights (or even random weights) won’t produce meaningful outputs.

To establish that connection, you’ll need to fine-tune both the image encoder and the decoder jointly. One of the key findings in our paper is that our model achieves the best performance among competing methods even in few-shot scenarios — when only a small number of slices are available for fine-tuning.

**Comment by sainatarajan** [comment]
@haoyudong-97 Thank you for providing me with the detailed information! I will mark the issue as solved.
